src/librustc_codegen_llvm/mir/operand.rs

```diff
@@ -85,12 +88,65 @@ impl OperandRef<'ll, 'tcx> {
             return Ok(OperandRef::new_zst(bx.cx, layout));
         }
 
-        match val.val {
+        let econv = |err| ConstEvalErr {
+            error: err,
+            stacktrace: Vec::new(),
+            span,
+        };
+
+        let val = match val.val {
             ConstValue::Unevaluated(..) => bug!(),
             ConstValue::ByRef(_, alloc, offset) => {
-                Ok(PlaceRef::from_const_alloc(bx, layout, alloc, offset).load(bx))
+                // FIXME: the first two arms are needed for simd_simple_float_intrinsic which reads
+                // the constants back from llvm values. We can probably do better.
+                match layout.abi {
+                    layout::Abi::Scalar(ref scalar) => {
+                        let x = alloc.read_scalar(
+                            bx.tcx(), offset, layout.size, layout.align,
+                        ).map_err(econv)?;
+                        let llval = scalar_to_llvm(
+                            bx.cx,
+                            x,
+                            scalar,
+                            layout.immediate_llvm_type(bx.cx),
+                        );
+                        OperandValue::Immediate(llval)
+                    },
+                    layout::Abi::ScalarPair(ref a_scalar, ref b_scalar) => {
```

Comment:
Wtf. This is a duplicate of... `read_value`, I guess. Why can't we use that instead?